File: python/data_structures/weighted_graph.py
from priority_q import PriorityQueue
class WeightedGraph:
	class edge:
		def __init__(self,node,weight):
			self.node = node
			self.weight = weight

		# repr is just for debug purposes.
		def __repr__(self):
			return f'{{ "node":{self.node},"weight":{self.weight} }}'

	def __init__(self):
		self.adj_lists = dict()
		self.vertexes = 0

	def __str__(self):
		return str(self.adj_lists)

	def add_vertex(self,vertex):
		if not self.adj_lists.get(vertex,False):
			self.adj_lists[vertex] = []
			self.vertexes +=1

	def add_edge(self,src,dest,weight):
		for edge in self.adj_lists[src]:
			if edge.node == dest:
				return
		self.adj_lists[src].append(self.edge(dest,weight))
		self.adj_lists[dest].append(self.edge(src,weight))

	def remove_edge(self,v1,v2):
		for idx in range(self.adj_lists[v1].length):
			if self.adj_lists[v1][idx].node == v2:
				del self.adj_lists[v1][idx]
				break
		for idx in enumerate(self.adj_lists[v1].length):
			if self.adj_lists[v1][idx].node == v2:
				del self.adj_lists[v1][idx]
				break


	def remove_vertex(self,v1):
		while self.adj_lists[v1]:
			tmp = self.adj_lists[v1][0]
			self.remove_edge(v1,tmp.node)
		self.adj_lists.pop(v1)
		self.vertexes -=1

	def dfs(self,start):
		stack = [start]
		visited = {start: 1}
		result = []
		while stack:
			current = stack.pop()
			result.append(current)
			for neighbor in self.adj_lists[current]:
				if not visited.get(neighbor):
					stack.append(neighbor)
					visited[neighbor] = 1
		return result


	def shortest_path(self,start,finish):
		nodes = PriorityQueue()
		distances = {}
		previous = {}
		# close enough to infinity for this purposes.
		infinity = 1<<64
		path = []
		smallest = None
		for vertex in self.adj_lists.keys():
			if vertex == start:
				distances[vertex] = 0
				nodes.enqueue(vertex,0)
			else:
				distances[vertex] = infinity
				nodes.enqueue(vertex,infinity)
			previous[vertex] = None

		while len(nodes):
			smallest = nodes.dequeue().value
			if smallest == finish:
				while previous[smallest]:
					path.append(smallest)
					smallest = previous[smallest]
				break
			if smallest or distances[smallest] != infinity:
				for neighbor in self.adj_lists[smallest]:
					candidate = distances[smallest] + neighbor.weight
					next_neighbor = neighbor.node
					if candidate < distances[next_neighbor]:
						distances[next_neighbor] = candidate
						previous[next_neighbor] = smallest
						nodes.enqueue(next_neighbor, candidate)

		path.reverse()
		return [smallest]+path


if __name__ == "__main__":
	graph = WeightedGraph()
	graph.add_vertex("a")
	graph.add_vertex("b")
	graph.add_vertex("c")
	graph.add_vertex("d")
	graph.add_vertex("e")
	graph.add_vertex("f")
	graph.add_edge("a","b", 4)
	graph.add_edge("a","c", 2)
	graph.add_edge("b","e", 3)
	graph.add_edge("c","d", 2)
	graph.add_edge("c","f", 4)
	graph.add_edge("d","e", 3)
	graph.add_edge("d","f", 1)
	graph.add_edge("e","f", 1)

	print(graph.shortest_path("a", "f"))

File: php/data_structures/weighted_graph.php
<?php
require_once "priority_q.php";
require_once "graph.php";
class _wg_edge{
	/**
	 * @var mixed
	 */
	public $node;
	/**
	 * @var numeric
	 */
	public $weight;
	public function __construct($value,$weight){
		$this->node = $value;
		$this->weight = $weight;
	}
}

class WeightedGraph extends Graph{

	public function add_edge($src,$dest,$weight=0){
		if(!array_key_exists($src,$this->adj_lists) || !array_key_exists($dest,$this->adj_lists)){
			return;
		}
		if(!in_array($src,$this->adj_lists[$src])){
			array_push($this->adj_lists[$src],new _wg_edge($dest,$weight));
			array_push($this->adj_lists[$dest],new _wg_edge($src,$weight));
		}
	}
	public function remove_edge($v1,$v2){
		$this->adj_lists[$v1] = array_values(array_filter($this->adj_lists[$v1],function($x) use ($v2) {
			return $x->node !== $v2;
		}));
		$this->adj_lists[$v2] = array_values(array_filter($this->adj_lists[$v2], function($x) use ($v1){
			return $x->node !== $v1;
		}));
	}
	public function remove_vertex($v1){
		while($this->adj_lists[$v1]){
			$v2 = $this->adj_lists[$v1][0]->node;
			$this->remove_edge($v1,$v2);
		}
		unset($this->adj_lists[$v1]);
	}
	public function dfs($start): array{
		$stack = new Stack();
		$visited = array();
		$stack->push($start);
		$visited[$start] = 1;
		$result = [];
		while($stack->size()){
			$current = $stack->pop();
			array_push($result,$current);
			foreach($this->adj_lists[$current] as $neighbor){
				if(!array_key_exists($neighbor->node,$visited)){
					$stack->push($neighbor->node);
					$visited[$neighbor->node] = 1;
				}
			}
		}
		return $result;
	}
	public function shortest_path($start,$finish): array {
		$nodes = new PriorityQueue();
		$distances = array();
		$previous = array();
		$inf = 1<<32;
		$path = array();
		$smallest = null;
		foreach(array_keys($this->adj_lists) as $vertex){
			if($start === $vertex){
				$distances[$vertex] = 0;
				$nodes->enqueue($vertex,0);
			}
			else{
				$distances[$vertex] = $inf;
				$nodes->enqueue($vertex,$inf);
			}
		}
		while($nodes->size()){
			$smallest = $nodes->dequeue()->value;
			if($smallest === $finish){
				while(array_key_exists($smallest,$previous) && $previous[$smallest]){
					array_push($path,$smallest);
					$smallest = $previous[$smallest];
				}
				break;
			}
			else if($smallest || $distances[$smallest] !== $inf){
				foreach($this->adj_lists[$smallest] as $neighbor){
					$candidate = $distances[$smallest] + $neighbor->weight;
					$next_neighbor = $neighbor->node;
					if($candidate < $distances[$next_neighbor]){
						$distances[$next_neighbor] = $candidate;
						$previous[$next_neighbor] = $smallest;
						$nodes->enqueue($next_neighbor, $candidate);
					}
				}
			}
		}
		array_push($path, $smallest);
		return array_reverse($path);
	}

}

if (!count(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS))){
	$g = new WeightedGraph();
	$g->add_vertex('a');
	$g->add_vertex('b');
	$g->add_vertex('c');
	$g->add_vertex('d');
	$g->add_vertex('e');
	$g->add_vertex('f');
	$g->add_edge('a', 'b', 4);
	$g->add_edge('a', 'c', 2);
	$g->add_edge('b', 'd', 3);
	$g->add_edge('c', 'd', 2);
	$g->add_edge('c', 'f', 4);
	$g->add_edge('d', 'e', 3);
	$g->add_edge('d', 'f', 1);
	$g->add_edge('e', 'f', 1);
	print_r($g);
	echo "shortest path PHP_EOL";
	print_r($g->shortest_path('a','f'));
	print_r($g->dfs('a'));
	$g->remove_vertex('b');
	print_r($g);
}


